### ISSUE TITLE
Should the process model be documented?

### ISSUE THREAD

**Issue by korydraughn** [phdavis1027/protocol_testing_env#2]
https://github.com/phdavis1027/protocol_testing_env/blob/2462b634d532b0567d09ef0958cd9bd30d11d145/the_protocol/notes.txt#L1-L5

We have this: https://docs.irods.org/4.3.0/system_overview/process_model/

Does that cover what you mentioned here?

Do you feel understanding the protocol requires some knowledge about the process model?

**Comment by phdavis1027** [comment]
Maybe this might clear up the simplified interface that the protocol presents? Something like, "every agent/process is just responsible for funneling requests to the state-changing part of the server, so after connection everything is a loop of requests and responses."

**Comment by korydraughn** [comment]
> Maybe this might clear up the simplified interface that the protocol presents?

Please elaborate. I sense you were confused by the protocol in some way?

> Something like, "every agent/process is just responsible for funneling requests to the state-changing part of the server, so after connection everything is a loop of requests and responses."

It seems what you are trying to capture is not the process model, but the message loop. The message loop is independent of the process model. The process model can change without affecting the message loop.

**Comment by phdavis1027** [comment]
> > Maybe this might clear up the simplified interface that the protocol presents?
> 
> Please elaborate. I sense you were confused by the protocol in some way?

I don't think so. Just that there are fewer entities from the protocol's perspective than there are from the perspective of iRODS as a whole. As far as the protocol is concerned, there is only one client and one server, with no differentiation between provider, consumer, and catalog. 

The process model might make it clear how this simplified client-server relationship can happen with several connections at once, where the underlying relationship between different nodes might be different. For example, nodes A and B are consumers, and node C is a provider. My understanding is that C can provide API responses to A while sending API requests to B. Under normal circumstances we might think of "provider" as meaning "server" and "consumer" as meaning "client," but because the session state is different for the processes associated with A and B, the protocol can  ignore those complexities. This is likely a pretty rough-grain understanding/explanation though...

> > Something like, "every agent/process is just responsible for funneling requests to the state-changing part of the server, so after connection everything is a loop of requests and responses."
> 
> It seems what you are trying to capture is not the process model, but the message loop. The message loop is independent of the process model. The process model can change without affecting the message loop.

You're totally right. I think the inspiration I'm drawing from the PG docs was that strictly speaking the process model is not relevant there either, but reveals just a little bit of implementation to make things more concrete for the reader.

It may even be that that is just a style choice, where some readers find that annoying and others like it.

**Comment by korydraughn** [comment]
I think you should write the spec in the way you feel makes things clear and easy to understand. Write it for the person who has never seen iRODS, but has to implement the protocol.

What do you feel that person wants to see in the spec?
What would you want to see in the spec?

Personally, I'd only include the things that are relevant to implementing the protocol, but that's just me. My point of view can be bias because I know a lot about the implementation.

If you feel including some information about the roles of a server leads to a clearer understanding of the protocol, then add it. We'll review and tweak as necessary.

There is no right or wrong answer here.
